--- ../src-base/minecraft/cpw/mods/fml/client/registry/RenderingRegistry.java
+++ ../src-work/minecraft/cpw/mods/fml/client/registry/RenderingRegistry.java
@@ -27,6 +27,8 @@
 import com.google.common.collect.Maps;
 import com.google.common.collect.ObjectArrays;
 
+import cpw.mods.fml.client.TextureFXManager;
+
 /**
  * @author cpw
  *
@@ -94,7 +96,41 @@
         return instance().nextRenderId++;
     }
 
+    /**
+     * Add a texture override for the given path and return the used index
+     *
+     * @param fileToOverride
+     * @param fileToAdd
+     */
+    @Deprecated
+    public static int addTextureOverride(String fileToOverride, String fileToAdd)
+    {
+        return -1;
+    }
 
+    /**
+     * Add a texture override for the given path and index
+     *
+     * @param path
+     * @param overlayPath
+     * @param index
+     */
+    public static void addTextureOverride(String path, String overlayPath, int index)
+    {
+//        TextureFXManager.instance().addNewTextureOverride(path, overlayPath, index);
+    }
+
+    /**
+     * Get and reserve a unique texture index for the supplied path
+     *
+     * @param path
+     */
+    @Deprecated
+    public static int getUniqueTextureIndex(String path)
+    {
+        return -1;
+    }
+
     @Deprecated public static RenderingRegistry instance()
     {
         return INSTANCE;
