--- ../src-base/minecraft/org/bukkit/GameMode.java
+++ ../src-work/minecraft/org/bukkit/GameMode.java
@@ -10,7 +10,13 @@
  * Represents the various type of game modes that {@link HumanEntity}s may have
  */
 public enum GameMode {
+    // Cauldron start - FakePlayers do not set their gametype and use the default
     /**
+     * Default mode
+     */
+    NOT_SET(-1),
+    // Cauldron end
+    /**
      * Creative mode may fly, build instantly, become invulnerable and create free items.
      */
     CREATIVE(1),
