--- ../src-base/minecraft/cpw/mods/fml/common/asm/transformers/deobf/FMLDeobfuscatingRemapper.java
+++ ../src-work/minecraft/cpw/mods/fml/common/asm/transformers/deobf/FMLDeobfuscatingRemapper.java
@@ -108,6 +108,12 @@
                 }
             }
             classNameBiMap = builder.build();
+            // Special case some mappings for modloader mods
+            mcpBuilder.put("BaseMod","net/minecraft/src/BaseMod");
+            mcpBuilder.put("ModLoader","net/minecraft/src/ModLoader");
+            mcpBuilder.put("EntityRendererProxy","net/minecraft/src/EntityRendererProxy");
+            mcpBuilder.put("MLProp","net/minecraft/src/MLProp");
+            mcpBuilder.put("TradeEntry","net/minecraft/src/TradeEntry");
             mcpNameBiMap = mcpBuilder.build();
         }
         catch (IOException ioe)
