--- ../src-base/minecraft/cpw/mods/fml/common/network/EntitySpawnPacket.java
+++ ../src-work/minecraft/cpw/mods/fml/common/network/EntitySpawnPacket.java
@@ -102,7 +102,7 @@
         {
             ent.getDataWatcher().writeWatchableObjects(dos);
         }
-        catch (IOException e)
+        catch (Throwable e) // MCPC+ - change IOException to Throwable for binary compatibility
         {
             // unpossible
         }
