--- ../src-base/minecraft/net/minecraft/block/BlockCrops.java
+++ ../src-work/minecraft/net/minecraft/block/BlockCrops.java
@@ -54,10 +54,9 @@
             {
                 float f = this.getGrowthRate(par1World, par2, par3, par4);
 
-                if (par5Random.nextInt((int)(25.0F / f) + 1) == 0)
+                if (par5Random.nextInt((int)(par1World.growthOdds / par1World.spigotConfig.wheatModifier * (25.0F / f)) + 1) == 0) // Spigot
                 {
-                    ++l;
-                    par1World.setBlockMetadataWithNotify(par2, par3, par4, l, 2);
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(par1World, par2, par3, par4, this.blockID, ++l); // CraftBukkit
                 }
             }
         }
