--- ../src-base/minecraft/net/minecraft/block/BlockReed.java
+++ ../src-work/minecraft/net/minecraft/block/BlockReed.java
@@ -40,9 +40,9 @@
             {
                 int i1 = par1World.getBlockMetadata(par2, par3, par4);
 
-                if (i1 == 15)
+                if (i1 >= (byte) range(3, (par1World.growthOdds / par1World.spigotConfig.caneModifier * 15) + 0.5F, 15)) // Spigot
                 {
-                    par1World.setBlock(par2, par3 + 1, par4, this.blockID);
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(par1World, par2, par3 + 1, par4, this.blockID, 0); // CraftBukkit
                     par1World.setBlockMetadataWithNotify(par2, par3, par4, 0, 4);
                 }
                 else
