--- ../src-base/minecraft/cpw/mods/fml/relauncher/IFMLLoadingPlugin.java
+++ ../src-work/minecraft/cpw/mods/fml/relauncher/IFMLLoadingPlugin.java
@@ -28,6 +28,13 @@
 public interface IFMLLoadingPlugin
 {
     /**
+     * Return a list of classes that implement the ILibrarySet interface
+     *
+     * @return a list of classes that implement the ILibrarySet interface
+     */
+    @Deprecated
+    String[] getLibraryRequestClass();
+    /**
      * Return a list of classes that implements the IClassTransformer interface
      * @return a list of classes that implements the IClassTransformer interface
      */
