--- ../src-base/minecraft/cpw/mods/fml/common/ModContainerFactory.java
+++ ../src-work/minecraft/cpw/mods/fml/common/ModContainerFactory.java
@@ -20,6 +20,7 @@
 import cpw.mods.fml.common.discovery.ModCandidate;
 import cpw.mods.fml.common.discovery.asm.ASMModParser;
 import cpw.mods.fml.common.discovery.asm.ModAnnotation;
+import cpw.mods.fml.common.modloader.ModLoaderModContainer;
 
 public class ModContainerFactory
 {
@@ -33,8 +34,8 @@
         String className = modParser.getASMType().getClassName();
         if (modParser.isBaseMod(container.getRememberedBaseMods()) && modClass.matcher(className).find())
         {
-            FMLLog.severe("Found a BaseMod type mod %s", className);
-            FMLLog.severe("This will not be loaded and will be ignored. ModLoader mechanisms are no longer available.");
+            FMLLog.fine("Identified a BaseMod type mod %s", className);
+            return new ModLoaderModContainer(className, modSource, modParser.getBaseModProperties());
         }
         else if (modClass.matcher(className).find())
         {
